File: api/v1/endpoints/analytics_cross_module.py
"""
API endpoints pour les analytics cross-module
"""

from typing import Dict, Any, Optional
from datetime import date
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from core.security import get_current_user, Permission, require_permissions
from db.database import get_db
from services.analytics_cross_module_service import CrossModuleAnalytics

router = APIRouter()

@router.get("/analytics/cross-module/unified")
async def get_unified_analytics(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Récupère les analytics unifiés de tous les modules.
    Inclut les corrélations et prédictions cross-module.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service.get_unified_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de la récupération des analytics: {str(e)}"
        )

@router.get("/analytics/cross-module/correlations")
async def get_correlations(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Récupère les corrélations entre les différents modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._analyze_correlations(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse des corrélations: {str(e)}"
        )

@router.get("/analytics/cross-module/predictions")
async def get_predictions(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Récupère les prédictions ML cross-module.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_ml_predictions(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de la récupération des prédictions: {str(e)}"
        )

@router.get("/analytics/cross-module/recommendations")
async def get_recommendations(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Récupère les recommandations cross-module basées sur ML.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._generate_cross_module_recommendations(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de la génération des recommandations: {str(e)}"
        )

@router.get("/analytics/cross-module/hr-impact")
async def get_hr_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact RH sur les autres modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_hr_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact RH: {str(e)}"
        )

@router.get("/analytics/cross-module/production-impact")
async def get_production_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact production sur les autres modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_production_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact production: {str(e)}"
        )

@router.get("/analytics/cross-module/finance-impact")
async def get_finance_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact finance sur les autres modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_finance_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact finance: {str(e)}"
        )

@router.get("/analytics/cross-module/inventory-impact")
async def get_inventory_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact inventaire sur les autres modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_inventory_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact inventaire: {str(e)}"
        )

@router.get("/analytics/cross-module/weather-impact")
async def get_weather_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact météo sur tous les modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_weather_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact météo: {str(e)}"
        )

@router.get("/analytics/cross-module/projects-impact")
async def get_projects_impact(
    date_debut: Optional[date] = Query(None, description="Date de début de l'analyse"),
    date_fin: Optional[date] = Query(None, description="Date de fin de l'analyse"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Analyse l'impact des projets sur tous les modules.
    """
    try:
        service = CrossModuleAnalytics(db)
        return await service._get_projects_analytics(date_debut, date_fin)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'analyse de l'impact des projets: {str(e)}"
        )

File: frontend/src/services/analytics_cross_module.ts
/**
 * Service pour les analytics cross-module
 */

import axios from 'axios';
import { format } from 'date-fns';
import type {
  CrossModuleAnalytics,
  ModuleCorrelations,
  ModulePredictions,
  CrossModuleRecommendation,
  HRAnalytics,
  ProductionAnalytics,
  FinanceAnalytics,
  InventoryAnalytics,
  AnalyseMétéo,
  ProjectsAnalytics
} from '../types/analytics_cross_module';

const BASE_URL = '/api/v1/analytics/cross-module';

interface AnalyticsParams {
  date_debut?: Date;
  date_fin?: Date;
}

/**
 * Formate les paramètres de date pour l'API
 */
const formatDateParams = (params?: AnalyticsParams) => {
  if (!params) return {};
  
  const formatted: Record<string, string> = {};
  if (params.date_debut) {
    formatted.date_debut = format(params.date_debut, 'yyyy-MM-dd');
  }
  if (params.date_fin) {
    formatted.date_fin = format(params.date_fin, 'yyyy-MM-dd');
  }
  return formatted;
};

/**
 * Service pour les analytics cross-module
 */
export const analyticsCrossModuleService = {
  /**
   * Récupère les analytics unifiés
   */
  async getUnifiedAnalytics(params?: AnalyticsParams): Promise<CrossModuleAnalytics> {
    const { data } = await axios.get<CrossModuleAnalytics>(
      `${BASE_URL}/unified`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère les corrélations entre modules
   */
  async getCorrelations(params?: AnalyticsParams): Promise<ModuleCorrelations> {
    const { data } = await axios.get<ModuleCorrelations>(
      `${BASE_URL}/correlations`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère les prédictions ML
   */
  async getPredictions(params?: AnalyticsParams): Promise<ModulePredictions> {
    const { data } = await axios.get<ModulePredictions>(
      `${BASE_URL}/predictions`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère les recommandations
   */
  async getRecommendations(params?: AnalyticsParams): Promise<CrossModuleRecommendation[]> {
    const { data } = await axios.get<CrossModuleRecommendation[]>(
      `${BASE_URL}/recommendations`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact RH
   */
  async getHRImpact(params?: AnalyticsParams): Promise<HRAnalytics> {
    const { data } = await axios.get<HRAnalytics>(
      `${BASE_URL}/hr-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact production
   */
  async getProductionImpact(params?: AnalyticsParams): Promise<ProductionAnalytics> {
    const { data } = await axios.get<ProductionAnalytics>(
      `${BASE_URL}/production-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact finance
   */
  async getFinanceImpact(params?: AnalyticsParams): Promise<FinanceAnalytics> {
    const { data } = await axios.get<FinanceAnalytics>(
      `${BASE_URL}/finance-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact inventaire
   */
  async getInventoryImpact(params?: AnalyticsParams): Promise<InventoryAnalytics> {
    const { data } = await axios.get<InventoryAnalytics>(
      `${BASE_URL}/inventory-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact météo
   */
  async getWeatherImpact(params?: AnalyticsParams): Promise<AnalyseMétéo> {
    const { data } = await axios.get<AnalyseMétéo>(
      `${BASE_URL}/weather-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  },

  /**
   * Récupère l'impact des projets
   */
  async getProjectsImpact(params?: AnalyticsParams): Promise<ProjectsAnalytics> {
    const { data } = await axios.get<ProjectsAnalytics>(
      `${BASE_URL}/projects-impact`,
      { params: formatDateParams(params) }
    );
    return data;
  }
};

/**
 * Hook React Query pour les analytics cross-module
 */
export const useAnalyticsCrossModule = () => {
  return {
    getUnifiedAnalytics: (params?: AnalyticsParams) => 
      analyticsCrossModuleService.getUnifiedAnalytics(params),
    getCorrelations: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getCorrelations(params),
    getPredictions: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getPredictions(params),
    getRecommendations: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getRecommendations(params),
    getHRImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getHRImpact(params),
    getProductionImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getProductionImpact(params),
    getFinanceImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getFinanceImpact(params),
    getInventoryImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getInventoryImpact(params),
    getWeatherImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getWeatherImpact(params),
    getProjectsImpact: (params?: AnalyticsParams) =>
      analyticsCrossModuleService.getProjectsImpact(params)
  };
};

export default analyticsCrossModuleService;
